beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.beam.direct.io;
+
+import com.google.common.base.Preconditions;
+import cz.o2.proxima.direct.commitlog.Offset;
+import java.io.Serializable;
+import java.util.Objects;
+import lombok.Getter;
+import lombok.ToString;
+import org.apache.beam.sdk.transforms.splittabledofn.HasDefaultTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.SplitResult;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A {@link RestrictionTracker} for {@link Offset Offsets} read from {@link
+ * cz.o2.proxima.direct.commitlog.CommitLogReader}
+ */
+public class OffsetRestrictionTracker
```

Comment:
couple of observations from going trough other `RestrictionTracker`
- We should keep `PositionT` immutable (as it is used as output element for split).
- Split (PositionT / Offset) must be either processed all at once, or we need to split it. Therefore RestrictionTracker can track current "position" within Offset. Finalization should be perfomed in `checkDone` method.
- We should get rid of all synchornizations as global locks are held in runner code.
- For get progress method, we should implement `HasProgress` interface

Reply:
I confused PositionT with `RestrictionT`. Can we keep RestrictionT immutable?

Reply:
No, we cannot. We have to move the restriction as we consume from unbounded restriction. Otherwise, we would have to recreate the restriction for each element.

Reply:
Why can't we track the state of "currently consumed restriction" inside `RestrictionTracker` (we consume only one restriction at the time and we need to the whole restriction or split it) and keep restriction itself stateless?

Am I missing something here?

Reply:
OK, `No, we cannot.` is not accurate. Correct is "No, we should not". The restriction *is* mutable, as it changes along the way we consume elements (if it is not pre-split, which is not the case here, but can be in general). We should implement things the way they are, not in an idiomatic way. That is to say - yes, we can threat the restriction as immutable, but it will not bring us anything, it will make the code more complicated and will not spare us from anything.

Reply:
Think more about this, as long as split always returns "deep copy", it should be ok

Reply:
Yes, that's definitely a requirement.